src/utils/stark/backwards-compatible/crypto.ts

```diff
@@ -0,0 +1,180 @@
+
+ https://github.com/starkware-libs/starkware-crypto-utils/blob/dev/src/js/key_derivation.js#L119
+*/
+export function grindKey(keySeed: BN): string {
+  const maxAllowedVal = MAX_ALLOWED_VAL();
+
+  // The key passed to hashKeyWithIndex must have a length of 64 characters
+  // to ensure that the correct number of leading zeroes are used as input
+  // to the hashing loop
+  let key = hashKeyWithIndex(keySeed.toString('hex', 64), 0);
+
+  // Make sure the produced key is devided by the Stark EC order, and falls within the range
+  // [0, maxAllowedVal).
+  for (let i = 0; key.gte(maxAllowedVal); i++) {
+    key = hashKeyWithIndex(key.toString('hex'), i);
+  }
+  return key.umod(starkEcOrder).toString('hex');
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////
+// DANGER: DO NOT MODIFY GRIND KEY / KEY GENERATION LOGIC.
+//////////////////////////////////////////////////////////////////////////////////////////////////
+/*
+The grindKeyV1 function introduced an alternate logic to the one in imx-sdk-js as part of this commit.
+https://github.com/immutable/imx-core-sdk/commit/3d9e35b4598784589bd7f121f26e105493196716
+
+Some of the accounts that has been created with versions released between 1.0.0-beta.3 and 2.0.0 may have an issue to regenerate their stark keys
+from the given l1signer. This function is retained to provide a way for those users to get their stark keys.
+
+For most users the above `grindKey` function should work correctly, only use `grindKeyV1` if the other one fails.
+
+Note: The issue we are addressing here is that if the hashKeyWithIndex value is above the limit, then we perform 
+      hash again with index until it comes below the limit. But for the first time the index is not incriminated in 
+      imx-sdk-js where as it was in core-sdk that was the difference. For this reason it would have worked for cases 
+      where the hashed value was less than the limit and fails only when it is above the limit.
+      Refer, https://immutable.atlassian.net/browse/DX-2167
+*/
+export function grindKeyV1(keySeed: BN) {
+  const maxAllowedVal = MAX_ALLOWED_VAL();
+  // The key passed to hashKeyWithIndex must have a length of 64 characters
+  // to ensure that the correct number of leading zeroes are used as input
+  // to the hashing loop
+  let key = hashKeyWithIndex(keySeed.toString('hex', 64), 0);
+
+  // Make sure the produced key is devided by the Stark EC order, and falls within the range
+  // [0, maxAllowedVal).
+  for (let i = 1; key.gte(maxAllowedVal); i++) {
+    key = hashKeyWithIndex(key.toString('hex'), i);
+  }
+  return key.umod(starkEcOrder).toString('hex');
+}
+
+/**
+ * Generates a new Stark private key
+ * @returns the private key as a hex string
+ */
+export function generateStarkPrivateKey(): string {
+  const keyPair = starkEc.genKeyPair();
+  // These are/will always be new accounts no need to check.
+  return grindKey(keyPair.getPrivate());
```

Comment:
Should this be `grindKeyV1` ?

Reply:
yup, done. Now `grindKeyV1` is the default `grindKey`